Create/edit/view project pages

MaraJade opened the pull request:
<!--
  This is a guideline for what a PR should look like.
  Feel free to modify it to fit your specific needs.

  Please list the issue this fixes in the PR
-->

fixes issue #29 
## Changes in this PR.

<!--
  Please include a list of all things this PR will accomplish
  Use the checkbox syntax to show what is done and what is yet to be completed.
  This gives context to the diff.
-->
- [X] Create projects page
- [x] Edit projects page
- [x] View projects page
## Testing this PR.

<!--
  Please include a list of explicit instructions for testing this PR.
  If the instructions are 'run `make test`' say that,
  If they are more complicated be thorough.
-->
1. Run `nosetests`
2. Start the docker container
3. View projects is at `/projects`
4. Create project is at `/projects/create`
5. Edit projects can be accessed via the `EDIT` link on view projects. To guarantee access to the edit page, login as the admin user
### Expected Output.

<!--
  Please insert either a description of what should happen when testing
  your PR or a code-block with terminal output.
-->

```
$ nosetests
........................................
----------------------------------------------------------------------
Ran 40 tests in 0.678s

OK
```

@osuosl/devs

MorganEPatch commented on app/views.py:
```diff
@@ -273,3 +273,109 @@ def admin():
+        }
+
+        if users:
+            project['users'] = re.split('\s?,\s?', users)
```
This only seems to add users to a project, without differentiating permissions levels. Is that all we want here?

pop commented on app/forms.py:
```diff
@@ -67,3 +67,18 @@ class GenerateReportForm(Form):
+
+
+class CreateProjectForm(Form):
+    uri = StringField('URI:')
```
I'm not sure where it would go but should there be some validation that this is a uri and not just a string?

pop commented on app/templates/view_projects.html:
```diff
@@ -0,0 +1,68 @@
+<script>
+function clearForm(form)
+{
+    for (i = 0; i < form.elements.length; i++)
```
`i` is uninitialized so this will not work.

MaraJade committed:
Attempting to fix the infinite commit problem

MaraJade committed:
Missed some things in the add

MaraJade committed:
Everything works again

MaraJade committed:
Flake8

pop commented:
@MaraJade If this is a PR to add some pages to the webapp, can you add those pages and how to access them to the PR testing instructions?

pop commented on app/util.py:
```diff
@@ -52,7 +52,7 @@ def is_logged_in():
 
 def error_message(array):
-    if type(array) is dict:
+    if array is dict:
```
I don't think this check does what you want it to do.

pop commented on app/util.py:
```diff
@@ -61,7 +61,7 @@ def error_message(array):
             # There was an error
             return True
-    elif type(array) is list:
+    elif array is list:
```
Again, that's not how type-checking works in python.

pop commented on app/views/create_project.py:
```diff
@@ -0,0 +1,74 @@
+
+    form.members.choices = usernames
+    form.managers.choices = usernames
+    form.spectators.choices = usernames
```
I don't understand how this works. Why would the members, managers, and spectators all be the same list of users?
If it does in fact work, add a comment as to _why/how_ it works.

pop commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
 Flask==0.10.1
-pymesync==0.1.5
+pymesync==0.1.6
```
This was bumped to 0.1.7 recently.

aaroncohen73 replied:
@ElijahCaine Yeah, but 0.1.7 has absolutely no changes from 0.1.6 :P

pop commented:
@MaraJade **Feedback:**
1. First off,  I think the _/projects_ table is off-by-one in how it's displaying data. the EDIT link should be the left-most entry and everything else should be shifted to the right by one.
![what it looks like in my browser](http://share.elijahcaine.me/projects.png)
1. When entering invalid data (like an invalid URI) the page claims that the project was created successfully, but upon further inspection it does not exist.
2. We could probably add a few more tests cases, but I can't really say with conviction what those tests cases would be.
3. Any notes I included in the code.

That said, this is actually pretty good. When I use it I think it works pretty well and mostly small stuff that should be easy enough to address.

**note** because of the size of this PR I think we should get at least three +1's. There's a lot of moving parts and more eyes should help address anything that might be broken before it's in develop.

MaraJade committed:
Addressed pops comments on the pr

aaroncohen73 commented on app/views/edit_project.py:
```diff
@@ -0,0 +1,115 @@
+        res = ts.update_project(project=project, slug=slug)
+
+        # TODO: Better error handling
+        if 'error' in res:
```
This should be updated to use `error_message()`

aaroncohen73 commented on app/views/view_projects.py:
```diff
@@ -0,0 +1,65 @@
+                           test=app.config['TESTING'],
+                           token=session['token'])
+
+    user = get_user(session['user']['username'])
```
This should be changed to `user = session['user']`

aaroncohen73 commented on app/views/view_projects.py:
```diff
@@ -0,0 +1,65 @@
+    user = get_user(session['user']['username'])
+    error_message(user)
+
+    projects = ts.get_projects()
```
This `get_projects()` should be removed because you're also calling `get_projects()` below

aaroncohen73 commented on app/views/view_projects.py:
```diff
@@ -0,0 +1,65 @@
+        if request.form['name']:
+            query['name'] = request.form['name']
+        if request.form['slugs']:
+            query['slugs'] = re.split('\s?,\s?', request.form['slugs'])
```
Why are you using `re.split()` instead of `request.form['slugs'].split(',')`?

aaroncohen73 commented on app/views/view_projects.py:
```diff
@@ -0,0 +1,65 @@
+        flash("Invalid form input")
+
+    projects = ts.get_projects(query_parameters=query)
+    error_message(projects)
```
This line through line 48 should be changed to something along the lines of

``` python
if error_message(projects):
    flash('Error')
    projects = []
```

aaroncohen73 commented on app/views/view_projects.py:
```diff
@@ -0,0 +1,65 @@
+        projects = []
+
+    for project in projects:
+        project['members'] = []
```
You don't have to do this, but you could cut out a lot of code here by doing something like

``` python
users = project['users']
project['members'] = [user for user, permissions in users.iteritems() if permissions['member']]
```

or

``` python
users = project['users']
project['members'] = filter(users.iteritems(), lambda user, permissions: permissions['member'])
```

(though the first one would be more version-compatible because Python 3 `filter()` doesn't return a list)

aaroncohen73 commented on app/views/create_project.py:
```diff
@@ -0,0 +1,83 @@
+        res = ts.create_project(project=project)
+        error_message(res)
+
+        flash("Project successfully submitted.")
```
This should only flash if `error_message()` returns False

aaroncohen73 commented on app/templates/view_times.html:
```diff
@@ -16,10 +16,7 @@
         }
     }
 
-    var oldBody = document.getElementsByTagName('tbody')[0];
```
This code should replace `form.submit()`. The reason it was replaced was because every time the form was cleared the app would query TimeSync for new times (which would take several seconds if there were lots of times on the server). Because of this, the behavior for filter forms was changed to not make any TimeSync queries until "Submit" is clicked.

This might be changed in the future (there's an issue open for request pagination on timesync-node [here](https://github.com/osuosl/timesync-node/issues/279)), but for now this is the best option we have for making `/times` usable. I believe that the behavior should be changed on `/projects` too in order to make it consistent across the app.

aaroncohen73 commented on app/templates/view_projects.html:
```diff
@@ -0,0 +1,71 @@
+            form.elements[i].value = "";
+        }
+    }
+    form.submit();
```
`form.submit()` should replaced with the code in view_times.html (see my below comment).

aaroncohen73 commented on app/views/create_project.py:
```diff
@@ -0,0 +1,83 @@
+            if field.data and field.name is not 'csrf_token':
+                if field.name == 'slugs':
+                    project['slugs'] = re.split('\s?,\s?', form.slugs.data)
+                elif field.name == 'members':
```
The code to turn the permissions lists into a permissions dictionary could probably be made less nested and complicated. Maybe it could look something like this:

``` python
users = set(request.form['members'] + request.form['managers'] + request.form['spectators'])
for user in users:
    user_permissions = {
        "member": user in request.form['members'],
        "manager": user in request.form['managers'],
        "spectator": user in request.form['spectators']
    }
    permissions[user] = user_permissions
```

Since it appears both here and in edit_project.py maybe it should also be factored out into a function inside util.py

aaroncohen73 commented on app/views/edit_project.py:
```diff
@@ -0,0 +1,115 @@
+                if field.name == 'slugs':
+                    slugs = form.slugs.data
+                    project['slugs'] = re.split('\s?,\s?', slugs)
+                elif field.name == 'members':
```
See my above comment about this code in create_project.py

aaroncohen73 commented on app/templates/select_project.html:
```diff
@@ -0,0 +1,10 @@
+{% extends "layout.html" %}
```
You should probably get rid of this because it isn't being used anywhere

aaroncohen73 commented:
@MaraJade I've left some feedback

Other than that and fixing the table alignment on `/projects`, I would like to see some more thorough tests written before I give this my +1.

MaraJade committed:
Merge branch 'develop' into marajade/project_page

Conflicts:
	app/forms.py
	app/templates/view_times.html
	requirements.txt

MaraJade committed:
Implemented all comments in the PR except for more tests

aaroncohen73 commented:
In my feedback before, I actually meant that you should get rid of the `select_project.html` file entirely because it isn't being used anywhere.

mathuin commented:
Checked out branch, built virtualenv, installed requirements, ran `nosetests`, every test failed with this error:  

```
ERROR: Failure: ImportStringError (import_string() failed for 'config'. Possible
 reasons are:

- missing __init__.py in a package;
- package or module path not included in sys.path;
- duplicated package or module name taking precedence in sys.path;
- missing module, class, function or variable;

Debugged import:

- 'config' not found.

Original exception:

ImportError: No module named config)
```

Was there something I was supposed to do before running the tests?

aaroncohen73 commented:
@mathuin The README says that you're supposed to copy `config.py.dist` to `config.py` and use that for configuration, but it is kind of hard to see it if you're not reading it close enough.

mathuin commented:
Ah, okay, thanks!

mathuin commented:
Copied `config.py.dist` to `config.py`, set `DEBUG` and `TESTING` to `True`, left other settings alone, `nosetests` passes with flying colors.  

Docker container came up with no trouble, view and create project pages are present, edit button is present on admin version of view project page.

The off-by-one header issue noted earlier in this ticket hasn't been addressed.

MaraJade committed:
Fixed header off-by-one problem

aaroncohen73 commented on app/views/login.py:
```diff
@@ -27,6 +27,7 @@ def login():
             status = token['status']
         # Else success, redirect to index page
         else:
+            print session
```
This line should probably be deleted

MaraJade committed:
Forgot to remove debugging print statement

aaroncohen73 commented:
Could you add links to these pages on `/` and `/admin/`?

aaroncohen73 commented:
The filter form on `/projects` isn't working for me. I'm entering things into the form and clicking "Submit" but the page just reloads with all of the projects instead of with the ones matching the filter parameters.

MaraJade committed:
View projects filter form works now, username needs pymesync though

Pymesync doesn't support filtering projects by username yet, but it should work
when that gets fixed

aaroncohen73 commented on app/templates/view_projects.html:
```diff
@@ -0,0 +1,67 @@
+                {% if admin %}
+                    <td><a href="{{ url_for('edit_project', project=p['slugs'][0]) }}">EDIT</a></td>
+                {% else %}
+                    {% for manager in p['managers'] %}
```
If you view the page as a project manager that isn't an admin, this will create a lot of unnecessary empty `<td>`s

[Example](http://imgur.com/h7cey4J) (logged in as "test" with password "test" on the staging server)

MaraJade committed:
I'm not sure why there were empty tags, but they're gone now

aaroncohen73 commented:
You can access `/projects/edit` for any project as any user (even if that user doesn't have permission to edit that project). Although the user can't _actually_ edit the project and an error is returned if they try to submit the form, they should be met with a 403 error if they try to edit projects they don't have permission for.

MaraJade committed:
Non-admins can't view edit page for projects they're not in

aaroncohen73 commented on app/views/edit_project.py:
```diff
@@ -27,6 +27,15 @@ def edit_project():
+        if not user['username'] in project['users'] and not user['site_admin']:
+            return "Permission denied", 403
+    elif not user['site_admin']:
+        print 'b'
```
You should probably delete these

MaraJade committed:
I keep forgetting about the print statements

aaroncohen73 commented:
+1

Kennric commented:
+1
